ExtraEnemyCustomization/Customizations/EnemyAbilities/Abilities/CloakAbility.cs

```diff
@@ -0,0 +1,209 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using UnityEngine;
+using EECustom.Extensions;
+using UnityEngine.Rendering;
+
+namespace EECustom.Customizations.EnemyAbilities.Abilities
+{
+    public sealed class CloakAbility : AbilityBase<CloakBehaviour>
+    {
+        public float CloakOpacity { get; set; } = 0.0f;
+        public float CloakDuration { get; set; } = 1.0f;
+        public float DecloakDuration { get; set; } = 1.0f;
+        public float DecloakAfterDelay { get; set; } = -1.0f;
+        public bool HideShadow { get; set; } = false;
+        public bool RequireTagForDetectionWhileCloaking { get; set; } = true;
+        public bool AllowEABAbilityWhileCloaking { get; set; } = true;
+
+        public override void OnAbilityLoaded()
+        {
+            CloakDuration = Mathf.Max(CloakDuration, 0.0f);
+            DecloakDuration = Mathf.Max(DecloakDuration, 0.0f);
+        }
+    }
+
+    public class CloakBehaviour : AbilityBehaviour<CloakAbility>
+    {
+        public override bool RunUpdateOnlyWhileExecuting => false;
+        public override bool AllowEABAbilityWhileExecuting => Ability.AllowEABAbilityWhileCloaking;
+        public override bool IsHostOnlyBehaviour => false;
+
+        private readonly List<Handler> _handlers = new();
+        private State _cloakState = State.None;
+        private float _startTime = 0.0f;
```

Comment:
You could probably use a time and duration instead of start/end time to avoid floating point errors.
e.g.
```cs
private float _time = 0.0f;
private float _duration = 0.0f;

private void Update()
{
    if (_time >= _duration) return;
    _time += Time.deltaTime;
}
```

Reply:
Changed as Suggestion